internal/service/lakeformation/opt_in_test.go

```diff
@@ -0,0 +1,407 @@
+						LFTagPolicy: &awstypes.LFTagPolicyResource{
+							ResourceType:   awstypes.ResourceType(rs.Primary.Attributes["resource_data.0.lf_tag_policy.0.resource_type"]),
+							CatalogId:      aws.String(rs.Primary.Attributes["resource_data.0.lf_tag_policy.0.catalog_id"]),
+							ExpressionName: aws.String(rs.Primary.Attributes["resource_data.0.lf_tag_policy.0.expression_name"]),
+						},
+					}
+				},
+				"resource_data.0.table.0.name": func(rs *terraform.ResourceState) *awstypes.Resource {
+					return &awstypes.Resource{
+						Table: &awstypes.TableResource{
+							Name:         aws.String(rs.Primary.Attributes["resource_data.0.table.0.name"]),
+							DatabaseName: aws.String(rs.Primary.Attributes["resource_data.0.table.0.database_name"]),
+						},
+					}
+				},
+				"resource_data.0.table_with_columns.0.name": func(rs *terraform.ResourceState) *awstypes.Resource {
+					return &awstypes.Resource{
+						TableWithColumns: &awstypes.TableWithColumnsResource{
+							Name:         aws.String(rs.Primary.Attributes["resource_data.0.table_with_columns.0.name"]),
+							DatabaseName: aws.String(rs.Primary.Attributes["resource_data.0.table_with_columns.0.database_name"]),
+						},
+					}
+				},
+			}
+
+			for path, constructor := range resourceConstructors {
+				if v, ok := rs.Primary.Attributes[path]; ok && v != "" {
+					in.Resource = constructor(rs)
+					break
+				}
+			}
+
+			_, err := tflakeformation.FindOptInByID(ctx, conn, principalID, in.Resource)
+			if err != nil {
+				if errs.IsAErrorMessageContains[*awstypes.AccessDeniedException](err, "Insufficient Lake Formation permission(s) on Catalog") {
+					return nil
+				}
+				if errs.IsAErrorMessageContains[*awstypes.EntityNotFoundException](err, "") {
+					return nil
+				}
+				return create.Error(names.LakeFormation, create.ErrActionCheckingDestroyed, tflakeformation.ResNameOptIn, rs.Primary.ID, err)
+			}
+			return create.Error(names.LakeFormation, create.ErrActionCheckingDestroyed, tflakeformation.ResNameOptIn, rs.Primary.ID, errors.New("not destroyed"))
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckOptInExists(ctx context.Context, name string, optin *lakeformation.ListLakeFormationOptInsOutput) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return create.Error(names.LakeFormation, create.ErrActionCheckingExistence, tflakeformation.ResNameOptIn, name, errors.New("not found"))
+		}
+
+		if rs.Primary.ID == "" {
+			return create.Error(names.LakeFormation, create.ErrActionCheckingExistence, tflakeformation.ResNameOptIn, name, errors.New("not set"))
+		}
+
+		conn := acctest.Provider.Meta().(*conns.AWSClient).LakeFormationClient(ctx)
+
+		principalID := rs.Primary.ID
+		in := &lakeformation.ListLakeFormationOptInsInput{}
+
+		type resourceConstructor func(*terraform.ResourceState) *awstypes.Resource
+
+		resourceConstructors := map[string]resourceConstructor{
```

Comment:
Could both of these be in a separate function so they can be shared by exists and destroy?